File: backend_flask/controllers/profile_controller.py
from flask import current_app, request

from services.profile_service import (
    get_profile_for_user,
    save_profile_avatar_for_user,
    update_profile_for_user,
)
from utils.response import error_response, success_response
from validators.profile_validator import validate_profile_payload


def _body():
    return request.get_json(silent=True) or {}


def _error_status(error):
    return getattr(error, "status_code", 500)


def get_profile():
    try:
        data = get_profile_for_user(request.user["id"])
        return success_response(200, "Profile fetched successfully.", data)
    except Exception as error:
        current_app.logger.exception("Get profile error")
        return error_response(
            _error_status(error),
            str(error) or "Failed to fetch profile.",
        )


def update_profile():
    payload = _body()

    try:
        validation_error = validate_profile_payload(payload)

        if validation_error:
            return error_response(400, validation_error)

        data = update_profile_for_user(request.user["id"], payload)
        return success_response(200, "Profile updated successfully.", data)
    except Exception as error:
        current_app.logger.exception("Update profile error")
        return error_response(
            _error_status(error),
            str(error) or "Failed to update profile.",
        )


def upload_profile_avatar():
    payload = _body()

    try:
        data = save_profile_avatar_for_user(
            request.user["id"],
            payload.get("imageData"),
        )
        return success_response(201, "Profile image uploaded successfully.", data)
    except Exception as error:
        current_app.logger.exception("Upload profile avatar error")
        return error_response(
            _error_status(error),
            str(error) or "Failed to upload profile image.",
        )

File: backend/src/controllers/profile.controller.js
import { saveProfileAvatarForUser } from "../services/profile/avatar.service.js";
import {
  getProfileForUser,
  getProfileSubmissionsForUser,
} from "../services/profile/read.service.js";
import { updateProfileForUser } from "../services/profile/update.service.js";
import { errorResponse, successResponse } from "../utils/response.js";
import { validateProfilePayload } from "../validators/profile.validator.js";

function getSubmissionListOptions(req) {
  return {
    page: req.query.page,
    limit: req.query.limit,
    filter: req.query.filter,
  };
}

export async function getProfile(req, res) {
  try {
    const data = await getProfileForUser(req.user.id);
    return successResponse(res, 200, "Profile fetched successfully.", data);
  } catch (error) {
    console.error("Get profile error:", error);
    return errorResponse(
      res,
      error.statusCode || 500,
      error.message || "Failed to fetch profile.",
    );
  }
}

export async function getProfileSubmissions(req, res) {
  try {
    const data = await getProfileSubmissionsForUser(
      req.user.id,
      getSubmissionListOptions(req),
    );
    return successResponse(res, 200, "Profile submissions fetched.", data);
  } catch (error) {
    console.error("Get profile submissions error:", error);
    return errorResponse(
      res,
      error.statusCode || 500,
      error.message || "Failed to fetch profile submissions.",
    );
  }
}

export async function updateProfile(req, res) {
  try {
    const validationError = validateProfilePayload(req.body); // check if payload is valid

    if (validationError) {
      return errorResponse(res, 400, validationError);
    }

    const data = await updateProfileForUser(req.user.id, req.body);
    return successResponse(res, 200, "Profile updated successfully.", data);
  } catch (error) {
    console.error("Update profile error:", error);
    return errorResponse(
      res,
      error.statusCode || 500,
      error.message || "Failed to update profile.",
    );
  }
}

export async function uploadProfileAvatar(req, res) {
  try {
    const data = await saveProfileAvatarForUser(
      req.user.id,
      req.body?.imageData,
    );

    return successResponse(
      res,
      201,
      "Profile image uploaded successfully.",
      data,
    );
  } catch (error) {
    console.error("Upload profile avatar error:", error);
    return errorResponse(
      res,
      error.statusCode || 500,
      error.message || "Failed to upload profile image.",
    );
  }
}
